distributed/scheduler.py

```diff
@@ -1590,6 +1596,36 @@ def get_nbytes(self, stream=None, keys=None, summary=True):
 
             return result
 
+    def publish_dataset(self, stream=None, keys=None, data=None, name=None,
+                        client=None):
+        for k in keys:
+            self.who_wants[k].add('published-%s' % name)
+        self.published_data[name] = {'data': data, 'keys': keys,
+                                     'clients': {client}}
+
+    def unpublish_dataset(self, stream=None, name=None):
+        out = self.published_data.pop(name, {'keys': []})
+        for k in out['keys']:
+            self.who_wants[k].remove('published-%s' % name)
```

Comment:
> Probably with the addition of unpublish, can now stop counting executors wanting a published dataset and regard each dataset as an ordinary client (without a stream), as you suggested some time before.

This does seem like a nice solution, and remove the need to think about this problem at all.